File: TestProjects/CPP_EchoTest/CPPEcho/CppEchoServer/TestNetObject.cpp
#include "TestNetObject.h"
#include "standard.h"

void TestNetObject::SetInfo()
{
	m_int64 = 12231126534198247U;
	m_float = 75.2165654f;
	m_wstr = L"hello world!!";
	integer_vec.push_back(10);
	integer_vec.push_back(90);
	integer_vec.push_back(70);
	integer_vec.push_back(50);
	integer_vec.push_back(44);
	int_to_wstr_map.insert({ 8,L"kdkdkdkdk" });
	int_to_wstr_map.insert({ 4,L"icicicicicicic oc" });
	int_to_wstr_map.insert({ 111,L"pipipipipipipip" });

	class_ptr = new TestMemeberClass();
	class_ptr->x = 1213.8f;
	class_ptr->y = -0.021f;
	class_ptr->z = 111.f;
}

void TestNetObject::Print() const
{
	printf("==== print\n");
	printf("%llu\n", m_int64);
	printf("%f\n", m_float);
	printf("%ws\n", m_wstr.c_str());

	printf("vector : ");
	for (const auto& item : integer_vec)
	{
		printf("%d ", item);
	}
	printf("\n");

	printf("map : ");
	for (const auto& item : int_to_wstr_map)
	{
		printf("[ %d , %ws ]\n", item.first, item.second.c_str());
	}
	printf("\n");

	printf("==== print end\n");
}

int TestNetObject::Serialize(std::basic_ostream<byte>& out_stream)
{
	int size = 0;	
	size += Utils::WriteToBinStream(out_stream, m_int64);
	size += Utils::WriteToBinStream(out_stream, m_float);
	size += Utils::WriteToBinStream(out_stream, m_wstr);
	size += Utils::WriteToBinStream(out_stream, integer_vec);
	size += Utils::WriteToBinStream(out_stream, int_to_wstr_map);
	size += Utils::WriteToBinStream(out_stream, class_ptr);
	return size;
}

int TestNetObject::DeSerialize(std::basic_istream<byte>& in_stream)
{
	int size = 0;	
	size += Utils::ReadFromBinStream(in_stream, m_int64);
	size += Utils::ReadFromBinStream(in_stream, m_float);
	size += Utils::ReadFromBinStream(in_stream, m_wstr);
	size += Utils::ReadFromBinStream(in_stream, integer_vec);
	size += Utils::ReadFromBinStream(in_stream, int_to_wstr_map);
	size += Utils::ReadFromBinStream(in_stream, class_ptr);
	return size;
}

int TestMemeberClass::Serialize(std::basic_ostream<byte>& out_stream)
{
	int size = 0;
	size += Utils::WriteToBinStream(out_stream, x);
	size += Utils::WriteToBinStream(out_stream, y);
	size += Utils::WriteToBinStream(out_stream, z);
	return size;
}

int TestMemeberClass::DeSerialize(std::basic_istream<byte>& in_stream)
{
	int size = 0;
	size += Utils::ReadFromBinStream(in_stream, x);
	size += Utils::ReadFromBinStream(in_stream, y);
	size += Utils::ReadFromBinStream(in_stream, z);
	return size;
}

void TestMemeberClass::Print() const
{
	printf("(%f,%f,%f)\n", x, y, z);
}

File: TestProjects/CPP_EchoTest/CPPEcho/CSEchoClient/TestNetObject.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;



namespace CSEchoClient
{
    interface ISerializable
    {
        abstract int Serialize(MemoryStream stream);
        abstract int DeSerialize(MemoryStream stream);
    }

    class TestMemberClass : ISerializable
    {
        float x;
        float y;
        float z;
        Int32 a;
        public TestMemberClass()
        {
            x = 123.123f;
            y = 0.4874f;
            z = 7154.454f;
            a = 1;
        }

        public int DeSerialize(MemoryStream stream)
        {
            int size = 0;
            size += MyConverter.ReadFromBinStream(stream, out x);
            size += MyConverter.ReadFromBinStream(stream, out y);
            size += MyConverter.ReadFromBinStream(stream, out z);
            size += MyConverter.ReadFromBinStream(stream, out a);
            return size;
        }

        public int Serialize(MemoryStream stream)
        {
            int size = 0;
            size += MyConverter.WriteToBinStream(stream, x);
            size += MyConverter.WriteToBinStream(stream, y);
            size += MyConverter.WriteToBinStream(stream, z);
            size += MyConverter.WriteToBinStream(stream, a);
            return size;
        }

        public void Print()
        {
            Console.WriteLine("{0}, {1}, {2}", x, y, z);
        }
    }

    class TestNetObject : ISerializable
    {
        UInt64 m_int64;
        float m_float;
        string m_str;
        List<Int32> integer_vec;
        Dictionary<int, string> int_to_wstr_map;
        TestMemberClass class_ptr;

        public TestNetObject()
        {
            class_ptr = new TestMemberClass();
        }

        public int DeSerialize(MemoryStream stream)
        {
            int size = 0;
            size += MyConverter.ReadFromBinStream(stream, out m_int64);
            size += MyConverter.ReadFromBinStream(stream, out m_float);
            size += MyConverter.ReadFromBinStream(stream, out m_str);
            size += MyConverter.ReadFromBinStream(stream, out integer_vec);
            size += MyConverter.ReadFromBinStream(stream, out int_to_wstr_map);
            size += MyConverter.ReadFromBinStream(stream, class_ptr);
            return size;
        }

        public int Serialize(MemoryStream stream)
        {
            int size = 0;            
            size += MyConverter.WriteToBinStream(stream, m_int64);
            size += MyConverter.WriteToBinStream(stream, m_float);
            size += MyConverter.WriteToBinStream(stream, m_str);
            size += MyConverter.WriteToBinStream(stream, integer_vec);
            size += MyConverter.WriteToBinStream(stream, int_to_wstr_map);
            size += MyConverter.WriteToBinStream(stream, class_ptr);
            return size;
        }

        public void SetInfo()
        {
            integer_vec = new List<int>();
            int_to_wstr_map = new Dictionary<int, string>();

            m_int64 = 12198247U;
            m_float = 0.200654f;
            m_str = "hello world!! this is client!!";
            integer_vec.Add(1022);
            integer_vec.Add(90456);
            integer_vec.Add(7078);
            integer_vec.Add(50445);
            integer_vec.Add(44232);
            int_to_wstr_map.Add(8, "kdkdkdkdk");
            int_to_wstr_map.Add(4, "icicicicicicic oc");
            int_to_wstr_map.Add(111, "pipipipipipipip");
        }

        public void Print()
        {
            Console.WriteLine("==== print");
            Console.WriteLine("{0}", m_int64);
            Console.WriteLine("{0}", m_float);
            Console.WriteLine("{0}", m_str);

            Console.WriteLine("vector : ");
            foreach (var item in integer_vec)
            {
                Console.WriteLine("{0} ", item);
            }
            Console.WriteLine();

            Console.WriteLine("map : ");
            foreach (var item in int_to_wstr_map)
            {
                Console.WriteLine("[ {0} , {1} ]", item.Key, item.Value);
            }
            Console.WriteLine();

            class_ptr.Print();

            Console.WriteLine("==== print end\n");
        }

    }



}
